enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,309 @@
+* specifies this service in log forwarding configuration
+* the service receives logs from a OKD cluster logging deployed collection agent  
+
+
+
+### Non-Goals
+
+Log forward intends to be opinionated about the source of log messages; it will explicitly define what constitutes "application logs" or "infrastructure logs". It will not allow configuration of additional sources without further design considerations and how those sources may impact the collecting agent. Log forwarding is not intended to provide a complex routing solution as one might achieve by using a custom collector configuration or a messaging solution (e.g. kafka).
+
+## Proposal
+
+Log forwarding will provide a declarative way to specify the destination for specific types of logs using a 'pipeline'.  A 'pipeline' defines simple routing for one source to one or more destinations. The source of logs are opinionated and well defined by cluster logging. The initial source types are as follows:
+
+* `logs.app` - Container logs generated by user applications running on the platform, excluding infrastructure container applications
+* `logs.infra` - Logs generated by both infrastructure components running on the platform and OKD nodes (e.g. journal logs).  "Infra" applications are defined as any pods which run in namespaces: `openshift*`, `kube*`, `default`.
+
+Future types may be:
+* `events.kubernetes` - Kubernetes events
+* `audit.container-engine` - Container engine audit events
+
+There are no assumptions regarding whether or not an endpoint is deployed on or off cluster.  Endpoints off-cluster may require adminstrators to perform additional actions in order for logs to be forwarded (e.g. secret creation, opening port) 
+Following is the list of supported endpoint types for this proposal:
+
+* Elasticsearch
+* Fluent secure-forward
+
+### User Stories
+
+#### As an OKD admin, I want to collect logs and send them to my own aggregation solution, for example Splunk
+
+This is a typical example of organizations that desires to re-use their existing enterprise log solution.  We will succeeded if we are able to write logs to their logging service.
+
+#### As an OKD admin, I want to aggregate application logs on-cluster and infra logs off-cluster
+
+This is an example of an OKD cluster hosting solution where several organizations are each provided with a dedicated cluster.  The organization requires access to application container logs but the host requires access to the infra structure logs.
+
+### Implementation Details
+
+#### Assumptions
+
+* A pipeline shall support at least one target endpoint. 
+* A pipeline may support multiple target endpoints.
+* Endpoints that require secrets which are not created by the cluster-logging-operator shall be created and managed by the administrator of the endpoint
+* Secrets shall have keys of: cert,key,cacert which point to the respective certificates for which they represent
```

Comment:
should we actually try to match more closely what is done for a `tls` type secret? They use `tls.crt` and `tls.key`.. we could also add `ca.crt` then

Reply:
So, for example, the steps for configuring secure_forward would be:
* user creates `secret/my-secure-forward-secret` which contains the "secret" parameters (that is, the parameters whose values we don't want in cleartext in the clusterlogging CR) used to configure https://docs.fluentd.org/output/forward:
```yaml
password: ASJDWEASJDFasd==
shared_key: KV203lnJSDFlasdf==
tls_cert_path: ASDJFASDFASDF==
tls_client_cert_path: XCKFWE2308945==
tls_client_private_key_path: ZXMCVOOEPWEP99==
etc.
```
well, the `path` parameters would be the actual contents of the cert/key files - the secret would mount these files into the pod, then we would generate fluentd config like
```
    tls_cert_path /etc/fluent/my-secure-forward-secret/tls_cert_path
```

Reply:
> should we actually try to match more closely what is done for a `tls` type secret? They use `tls.crt` and `tls.key`.. we could also add `ca.crt` then

This choice was made based on available options to `curl` but I'm open to values that are more consistent to what you may find elsewhere.  Should it be `tls.cacrt`  or `tls.ca.crt` or just `ca.crt`

Reply:
ok - I think using the `curl` names is fine - and for parameters for which there is no curl name e.g. `shared_key` we should use the name from the `fluentd` configuration.

Reply:
I think `ca.crt` makes the most sense since it would have been used to sign `tls.crt` and may not be exclusive to the cert provided in the secret

Reply:
Doesn't seem we have consensus here? More Opinions?

Reply:
The only thing I would request here is to be able to configure the TLS protocol version and the cipher suite. These options are very often requested by customers that need to meet regulations or compliance requirements.

Reply:
@JAORMX these items don't necessarily need to be in the secret which hides certs do they?  Seems the target endpoint config is a better place to optionally allow these items

Reply:
They do not need to be in the secret.  The only things that need to be in the secret are things which are either very large (like a cert) or things that we don't want in cleartext in the CR (like a key, password, etc.).

I think we may want to expose most of the TLS parameters in the CR - tls_version, tls_ciphers, etc.  I also notice there are Windows specific ones we may need to add for Windows support.  https://docs.fluentd.org/output/forward#tls_version

Note that the elasticsearch output plugin does not support specifying the cipher suites: https://github.com/uken/fluent-plugin-elasticsearch/blob/master/lib/fluent/plugin/out_elasticsearch.rb#L91 - there are very few tls parameters